File: 03-shiny-chat-app.R
library(dotenv) # Will read OPENAI_API_KEY from .env file
library(elmer)
library(shiny)
library(shinychat)

ui <- bslib::page_fluid(
  chat_ui("chat")
)

server <- function(input, output, session) {
  chat <- chat_openai(
    model = "gpt-4o",
    system_prompt = "You're a trickster who answers in riddles"
  )

  observeEvent(input$chat_user_input, {
    stream <- chat$stream_async(input$chat_user_input)
    chat_append("chat", stream)
  })
}

shinyApp(ui, server)

File: 03-shiny-chat-app.py
# To run this app, execute `shiny run 03-shiny-chat-app.py` at the terminal,
# or in Positron/VS Code with the Shiny extension, use the Run button-menu's
# "Run Shiny App" option.

from chatlas import ChatAnthropic
from dotenv import load_dotenv
from shiny.express import input, ui

_ = load_dotenv()

chat_session = ChatOpenAI(
    model="gpt-4.1",
    system_prompt="You're a trickster who answers in riddles",
)

chat = ui.Chat(id="chat")

chat.ui()


@chat.on_user_submit
async def handle_input():
    user_input = input.chat_user_input()
    response = await chat_session.stream_async(user_input)
    await chat.append_message_stream(response)
